### DISCUSSION TITLE
chap6-1 질문 정리

### DISCUSSION THREAD

**Discussion by sihyung92** [JS-OS-study/hyokyung-and-students#17] [General]
chap6, Process synchronization

**Comment by inseo24** [comment]
<h3>느낀 점 & 궁금한 점 </h3>

- 메모리를 공유하면서 발생할 수 있는 Process Synchronization 문제에 대해서 알게 되었다.
- race condition이 발생할 수 있는 상황을 보면서 많은 고민들로 문제가 해결되어 왔음을 느낌
- Test and Set과 세마포어?에 대해 잘 이해하지 못했다. 나중에 다시 들어야 할 거 같다. 

<hr />
<h3>문제</h3>

Q. 계속 CPU와 메모리를 쓰면서 기다리는 상황을 뭐라할까요?
A. ________ ( = _______ )

**Reply by pullingoff** [reply]
busy wait = spin lock

**Reply by 1chz** [reply]
Busy Wait (=Spin Lock)

**Reply by sihyung92** [reply]
Busy Wait = spin lock

**Comment by sihyung92** [comment]
# 느낀점 & 궁금한점
- race condition의 설명을 들으니, 코드를 짜다보면 대부분의 골치아픈 문제는  race condition에 기인한다는 생각이 든다.
- busy wating에 대해 설명해주셨는데 하드웨어 적인 방법이 도입되어도 busy waiting은 여전히 존재할 것 같다. 어떤 방식으로 해결할 수 있을까?

# 퀴즈
Q. 다음 중 옳지 않은 진술과 그 이유를 적어주세요!
1. 멀티 프로세서 환경에서 CPU 2개가 동일한 memory 접근에서 생기는 race condition을 방지하기 위해선, 커널에 1개의 CPU만 접근할 수 있도록 하는 방식이 효율적이다. (O/X)
2. 시스템 콜을 통해 커널모드에 진입되어 있는 프로세스는 선점할 수 없다. (O/X)
3. Critical Section은 하드웨어적으로 처리하는 것이 간편하므로 코드에선 신경쓰지 않아도 된다. (O/X)

**Reply by inseo24** [reply]
1. x
2. o 
3. x

**Reply by 1chz** [reply]
1. X
2. O
3. X

**Reply by pullingoff** [reply]
1 X 이유: 커널에 1개의 cpu만 접근할 수 있도록 하는 방식은 어쩔수 없지만 효율적이지 않음
2 O
3 X 이유: 하드웨어, 코드 둘다에서 신경써야함

**Comment by 1chz** [comment]
# 느낀점

---

- 불변성을 보장해준다면 이러한 문제를 신경쓰지 않아도 되겠다는 생각이 들었다.
- 대체로 데이터베이스에서 발생하는 문제와 매우 유사한 것들이라는 생각이 들었다.

# 문제

---

- `Deadlock`과 `Starvation`의 차이에 대해 서술하시오.

**Reply by pullingoff** [reply]
데드락은 프로세스끼리 서로 자원을 가져오려고 할때 발생하고 Starvation은 프로세스가 아무 자원도 갖고 있지 않을 때를 말한다.(?)

**Reply by inseo24** [reply]
데드락은 각 프로세스가 자원을 갖고 있는 상태에서 기다림
starvation은 자원 자체를 얻고 있지 않은 상황

**Reply by sihyung92** [reply]
Deadlock은 두개의 프로세스가 서로의 자원을 필요로하여 둘다 못갖는 상황 ㅠ이고, Starvation은 하나의 프로세스가 필요로 하는 자원에 접근을 못하는 것을 의미한다

**Comment by pullingoff** [comment]
# 느낀 점
교수님이 Block/wakeup P연산과 V연산에서 S의 의미가 좀 다르다고 했는데 이해가 잘 안된다. 좀 더 찾아보기

# 문제
1. 0 또는 1의 값만 가질 수 있는 세마포어를 한 단어로 뭐라고 하나요?
2. ![문제 2번](https://user-images.githubusercontent.com/50111853/154952318-6d2ac8c4-9604-4719-8b5d-797cd5dda2d0.png) 
사진 속 알고리즘은 critical section 알고리즘의 충족 조건 중 무엇을 충족하지 못했나요?

**Reply by inseo24** [reply]
1. mutex
2. progress

**Reply by 1chz** [reply]
1. Binary Semaphore
2. Progress

**Reply by sihyung92** [reply]
1. mutex
2. Progress
